### DISCUSSION TITLE
Improving reproducibility: Moving to UV

### DISCUSSION THREAD

**Discussion by KennethEnevoldsen** [embeddings-benchmark/mteb#1280] [Ideas]
We are already using `ruff` (instead of flake8 and/or black) it might similarly be great to move to [uv](https://docs.astral.sh/uv/). Uv is made by the same (great) team. 

A benefit would be that we can add a `.lock`-file to allow for more reproducibility. This will also make some of the CI easier going forward. 

`uv` is pip compatible so it would still be possible to use `pip` without issues (as far as I am aware).

To start the discussion I will just add a few people:
@isaac-chung, @orionw, @Muennighoff, @x-tabdeveloping

**Comment by KennethEnevoldsen** [comment]
(@MartinBernstorff might also have a few comments to add)

**Comment by Samoed** [comment]
`uv` is a good tool, but it doesn't support development dependency groups, which is a minor issue in my opinion. I use `pdm`, which has a similar CLI to `uv` and [can use `uv` as a backend, though this is experimental](https://pdm-project.org/latest/usage/uv/).

**Reply by KennethEnevoldsen** [reply]
are you referring to stuff like `pip install mteb[dev]`?

**Reply by mrshu** [reply]
@KennethEnevoldsen it is a similar concept, although what you refer to is called "optional dependencies", as far as I know.

This tracking GH issue has a bit more detail: https://github.com/astral-sh/uv/issues/5632

I would indeed call it a very minor issue.

**Reply by Samoed** [reply]
Yes, @mrshu right. [PDM info](https://pdm-project.org/latest/usage/dependency/#add-development-only-dependencies)

**Reply by KennethEnevoldsen** [reply]
Thanks for clarifying seems like we can at least keep the current setup for now (without loosing anything) and potentially switch to dependency groups going forward

**Comment by orionw** [comment]
As someone who has never used ‘uv’ I would just suggest that we have very good documentation for it if we do switch.  

My biggest hesitation is making it harder for new collaborators to have to learn a new package manager. I’ve heard great things about it, but TBF I had heard that about poetry and it was fairly non-intuitive to use, whereas everyone knows how to use pip.  But I haven’t used uv so perhaps it actually is easy :)

Edit: I see that it is pip compatible so maybe not an issue.

**Reply by KennethEnevoldsen** [reply]
Yea while poetry requires a complete redo of the pyproject.toml, `uv` simply adds a `uv.lock` and otherwise I don't believe we need to see a lot of changes.

One benefit of the uv.lock is that people would be able to see which exact versions people have got things to run with at a specific time.
